File: android/src/main/java/org/godotengine/plugin/firebase/FirebasePlugin.java
//
// © 2026-present https://github.com/firebase-team
//

package org.godotengine.plugin.firebase;

import android.app.Activity;
import android.util.Log;
import android.view.View;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.godotengine.godot.Godot;
import org.godotengine.godot.Dictionary;
import org.godotengine.godot.plugin.GodotPlugin;
import org.godotengine.godot.plugin.SignalInfo;
import org.godotengine.godot.plugin.UsedByGodot;


public class FirebasePlugin extends GodotPlugin {
	public static final String CLASS_NAME = FirebasePlugin.class.getSimpleName();
	static final String LOG_TAG = "godot::" + CLASS_NAME;


	static final String TEMPLATE_READY_SIGNAL = "template_ready";
	// TODO: Define all signals

	public FirebasePlugin(Godot godot) {
		super(godot);
	}

	@Override
	public String getPluginName() {
		return CLASS_NAME;
	}

	@Override
	public Set<SignalInfo> getPluginSignals() {
		Set<SignalInfo> signals = new HashSet<>();
		signals.add(new SignalInfo(TEMPLATE_READY_SIGNAL, Dictionary.class));

		// TODO: Register all signals

		return signals;
	}

	@Override
	public View onMainCreate(Activity activity) {
		// TODO: Godot activity is ready

		return super.onMainCreate(activity);
	}

	@Override
	public void onGodotSetupCompleted() {
		super.onGodotSetupCompleted();

		// TODO: Godot is ready
	}

	@UsedByGodot
	public Object[] get_firebase() {
		Log.d(LOG_TAG, "get_firebase() invoked");

		List<Dictionary> resultList = new ArrayList<>();

		// TODO: Plugin method

		return resultList.toArray();
	}

	@Override
	public void onMainDestroy() {
		// TODO: Plugin cleanup
	}
}

File: android/src/main/java/org/godotengine/plugin/firebase/FirebasePlugin.kt
//
// © 2026-present Firebase Team https://github.com/firebase-team
//
package org.godotengine.plugin.firebase

import android.app.Activity
import android.content.Intent
import android.util.Log
import android.view.View
import org.godotengine.godot.Dictionary
import org.godotengine.godot.Godot
import org.godotengine.godot.plugin.GodotPlugin
import org.godotengine.godot.plugin.SignalInfo
import org.godotengine.godot.plugin.UsedByGodot

import org.godotengine.plugin.firebase.model.FirestoreDocument


class FirebasePlugin(
    godot: Godot,
) : GodotPlugin(godot) {
    companion object {
        private const val TAG = "FirebasePlugin"
    }

    override fun getPluginName(): String = TAG

    private val auth = Authentication(this)
    private val firestore = Firestore(this)

    override fun onMainCreate(activity: Activity?): View? {
        val act = activity ?: getActivity()
        act?.let { auth.init(it) }
        return super.onMainCreate(activity)
    }

    override fun onMainActivityResult(
        requestCode: Int,
        resultCode: Int,
        intent: Intent?,
    ) {
        auth.handleActivityResult(requestCode, resultCode, intent)
    }

    override fun getPluginSignals(): MutableSet<SignalInfo> {
        val signals: MutableSet<SignalInfo> = mutableSetOf()
        signals.addAll(auth.authSignals())
        signals.addAll(firestore.firestoreSignals())
        Log.d(TAG, "Registering ${signals.size} signals")
        return signals
    }

    fun emitGodotSignal(signalName: String, vararg args: Any?) {
        getGodot().runOnRenderThread {
            emitSignal(signalName, *args)
        }
    }


    /**
     * Authentication
     */

    @UsedByGodot
    fun create_user(email: String, password: String) = auth.createUser(email, password)

    @UsedByGodot
    fun link_anonymous_with_google() = auth.linkAnonymousWithGoogle()

    @UsedByGodot
    fun sign_in(
        email: String,
        password: String,
    ) = auth.signIn(email, password)

    @UsedByGodot
    fun sign_in_with_google() = auth.signInWithGoogle()

    @UsedByGodot
    fun sign_in_anonymously() = auth.signInAnonymously()

    @UsedByGodot
    fun is_signed_in() = auth.isSignedIn()

    @UsedByGodot
    fun sign_out() = auth.signOut()

    @UsedByGodot
    fun send_verification_email() = auth.sendVerificationEmail()

    @UsedByGodot
    fun send_password_reset_email(email: String) = auth.sendPasswordResetEmail(email)

    @UsedByGodot
    fun get_current_user() = auth.getCurrentUser()

    @UsedByGodot
    fun delete_current_user() = auth.deleteCurrentUser()


	/**
	 * Firestore
	 */

	@UsedByGodot
	fun add_document(document: Dictionary) = firestore.addDocument(FirestoreDocument(document))

	@UsedByGodot
	fun set_document(document: Dictionary, merge: Boolean = false) =
            firestore.setDocument(FirestoreDocument(document), merge)

	@UsedByGodot
	fun get_document(collection: String, documentId: String) = firestore.getDocument(collection, documentId)

	@UsedByGodot
	fun update_document(document: Dictionary) = firestore.updateDocument(FirestoreDocument(document))

	@UsedByGodot
	fun delete_document(collection: String, documentId: String) = firestore.deleteDocument(collection, documentId)

	@UsedByGodot
	fun get_collection(collection: String) = firestore.getCollection(collection)

	@UsedByGodot
	fun track_document(collection: String, documentId: String) = firestore.trackDocument(collection, documentId)

	@UsedByGodot
	fun stop_tracking_document(collection: String, documentId: String) =
            firestore.stopTrackingDocument(collection, documentId)
}
